src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -518,5 +577,88 @@ internal JsonArrayPool(ArrayPool<T> arrayPool)
 
             public void Return(T[] array) => this.arrayPool.Return(array);
         }
+
+        private class JsonProgressClientConverter : JsonConverter
+        {
+            private JsonMessageFormatter formatter = null;
+
+            public JsonProgressClientConverter(JsonMessageFormatter formatter)
+            {
+                this.formatter = formatter;
+            }
+
+            public override bool CanConvert(Type objectType) => MessageFormatterHelper.FindIProgressOfT(objectType) != null;
+
+            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+            {
+                throw new NotSupportedException();
+            }
+
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                // if the requestId is empty it means the Progress object comes from a response
+                if (this.formatter.requestId == null)
+                {
+                    writer.WriteValue(value);
+                    return;
+                }
+
+                long progressId = Interlocked.Increment(ref this.formatter.nextProgressId);
+                this.formatter.requestProgressMap.Add(this.formatter.requestId, progressId);
+
+                this.formatter.progressMap.Add(progressId, value);
+
```

Comment:
You're going to need to acquire a lock for accessing these so as to not race with your code that accesses and removes from these maps.
And while you're at it, you can probably replace your Interlocked call with just a simple increment that exists in the lock as well.